src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,201 @@
+                easy.SetCurlOption(CURLoption.CURLOPT_SSLCERT, _certificateFile);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSLCERTTYPE, CertificateFileType);
+                easy.SetCurlOption(CURLoption.CURLOPT_KEYPASSWD, password);
+                VerboseTrace(_certificateFile + " created");
+            }
+
+            internal void SetClientCertificateOptions(EasyRequest easy)
+            {
+                Debug.Assert(s_clientCertificatePolicy == ClientCertificatePolicy.Callback ||
+                             s_clientCertificatePolicy == ClientCertificatePolicy.CommonName ||
+                             s_clientCertificatePolicy == ClientCertificatePolicy.FileExport);
+
+                _certificateFile = null;
+                switch (s_clientCertificatePolicy)
+                {
+                    case ClientCertificatePolicy.Callback:
+                        {
+                            SetCallbacks(easy);
+                            break;
+                        }
+                    case ClientCertificatePolicy.CommonName :
+                        {
+                            SetClientCertCommonName(easy);
+                            break;
+                        }
+                    default:
+                        ExportToFile(easy);
+                        break;
+                }
+            }
+
+            public void Dispose()
+            {
+                if (_certificateFile == null)
+                {
+                    return;
+                }
+                try
+                {
+                    File.Delete(_certificateFile);
+                }
+                catch (IOException e)
+                {
+                    VerboseTrace("Deleting temporary file:" + e);
+                }
+            }
+
+            internal static ClientCertProvider Create(ClientCertificateOption option, string uriScheme)
+            {
+                if (option == ClientCertificateOption.Manual || !string.Equals(uriScheme, UriSchemeHttps))
+                {
+                    return null;
+                }
+
+                using (var myStore = new X509Store())
+                {
+                    myStore.Open(OpenFlags.ReadOnly);
+                    X509Certificate2Collection clientCerts = myStore.Certificates.FindTlsClientCertificates();
+                    if (clientCerts.Count == 0)
+                    {
+                        VerboseTrace("0 Client Certs");
+                        return null; 
+                    }
+                    VerboseTrace("Client Certs found " + clientCerts.Count);
+
+                    // NOTE: This is a DEVIATION FROM WinHttpHandler
+                    // Unlike WinHttpHandler, we are going to use only one client certificate.
+                    // libcurl, libssl and even X509Certificate2Collection.Export all are able to handle single client certificate.
+                    // Picking the first available cert
```

Comment:
My guess is that on Unix, it would be rare to provision multiple certs on a machine.
If libcurl had exposed the underlying SSL connection, we could have done better filtering based on the optional CA list sent by the server. In the absence of that, I tend to agree with @kapilash that exposing a public delegate from CurlHandler is a cleaner solution.